core/sequencer/sequencer.go

```diff
@@ -124,49 +125,52 @@ func (s *Sequencer) Initialize(ctx context.Context, logID, mapID int64) error {
 // StartSequencingAll starts sequencing processes for all domains.
 func (s *Sequencer) StartSequencingAll(ctx context.Context, refresh time.Duration) error {
 	started := make(map[string]bool)
+	recievers := make(map[string]mutator.Reciever)
 	ticker := time.NewTicker(refresh)
 	defer func() { ticker.Stop() }()
 
-	for range ticker.C {
-		domains, err := s.domains.List(ctx, false)
-		if err != nil {
-			return fmt.Errorf("admin.List(): %v", err)
-		}
-		for _, d := range domains {
-			if !started[d.Domain] {
-				glog.Infof("StartSigning domain: %v", d.Domain)
-				started[d.Domain] = true
-				go s.StartSigning(ctx, d.LogID, d.MapID, d.MinInterval, d.MaxInterval)
+	for {
+		select {
+		case <-ticker.C:
+			domains, err := s.domains.List(ctx, false)
+			if err != nil {
+				return fmt.Errorf("admin.List(): %v", err)
+			}
+			for _, d := range domains {
+				if !started[d.Domain] {
+					glog.Infof("StartSigning domain: %v", d.Domain)
+					started[d.Domain] = true
+					recievers[d.Domain] = s.NewReciever(ctx, d.LogID, d.MapID, d.MinInterval, d.MaxInterval)
+				}
 			}
+		case <-ctx.Done():
+			for _, r := range recievers {
+				r.Close()
+			}
+			return ctx.Err()
 		}
 	}
-	return nil
 }
 
-// StartSigning advance epochs once per minInterval, if there were mutations,
-// and at least once per maxElapsed minIntervals.
-func (s *Sequencer) StartSigning(ctx context.Context, logID, mapID int64, minInterval, maxInterval time.Duration) {
-	if minInterval > maxInterval {
-		glog.Errorf("maxInterval: %v, want < minInterval: %v", maxInterval, minInterval)
-		return
-	}
-
+// NewReciever creates a new reciever for mapID.
+// New epochs will be created at least once per maxInterval and as often as minInterval.
+func (s *Sequencer) NewReciever(ctx context.Context, logID, mapID int64, minInterval, maxInterval time.Duration) mutator.Reciever {
 	if err := s.Initialize(ctx, logID, mapID); err != nil {
 		glog.Errorf("Initialize() failed: %v", err)
 	}
 	var rootResp *trillian.GetSignedMapRootResponse
-	ctxTime, cancel := context.WithTimeout(ctx, minInterval)
-	rootResp, err := s.tmap.GetSignedMapRoot(ctxTime, &trillian.GetSignedMapRootRequest{
+	cctx, cancel := context.WithTimeout(ctx, minInterval)
+	rootResp, err := s.tmap.GetSignedMapRoot(cctx, &trillian.GetSignedMapRootRequest{
 		MapId: mapID,
 	})
 	if err != nil {
 		glog.Infof("GetSignedMapRoot failed: %v", err)
 		// Immediately create new epoch and write new sth:
-		if err := s.CreateEpoch(ctxTime, logID, mapID, ForceNewEpoch(true)); err != nil {
+		if err := s.CreateEpoch(cctx, logID, mapID, nil); err != nil {
```

Comment:
This is the case where an empty epoch is created, right? (arg 4 = `nil`)
It'd be clearer to have `empty` here, defined somewhere as
```
empty := []*mutator.QueueMessage{}
```

Reply:
right.